### PR TITLE
[Merged by Bors] - feat(SetTheory/ZFC/Ordinal): Alternate characterizations of ordinals

### PR DESCRIPTION
**Pull request by vihdzp**
We previously defined ordinals as transitive sets, transitive under membership. This PR establishes this definition is equivalent to the following:
- An ordinal is a transitive set of transitive sets
- An ordinal is a transitive set of ordinals
- An ordinal is a transitive set, trichotomous under membership
- An ordinal is a transitive set, well-ordered under membership

---

- [x] depends on: #15793

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/SetTheory/ZFC/Ordinal.lean [hunk 1]
```diff
@@ -12,27 +13,24 @@ import Mathlib.SetTheory.ZFC.Basic
...
-We currently only have an initial development of transitive sets and ordinals. Further development
-can be found on the Mathlib 3 branch `von_neumann_v2`.
-
 ## Definitions
 
 - `ZFSet.IsTransitive` means that every element of a set is a subset.
 - `ZFSet.IsOrdinal` means that the set is transitive and well-ordered under `∈`.
 
-## TODO
+## Todo
```
Comment (alreadydone) [inline Mathlib/SetTheory/ZFC/Ordinal.lean:21]: I think TODO is the more common spelling in mathlib. To-do is also good to me.

File: Mathlib/SetTheory/ZFC/Ordinal.lean [hunk 2]
```diff
@@ -120,17 +120,130 @@ theorem mem_trans (h : z.IsOrdinal) : x ∈ y → y ∈ z → x ∈ z :=
...
+    rintro ⟨h₁, ⟨h₂⟩⟩
+    refine ⟨h₁, @fun y z w hyz hzw hwx ↦ ?_⟩
+    have hzx := h₁.mem_trans hzw hwx
+    exact h₂ ⟨y, h₁.mem_trans hyz hzx⟩ ⟨z, hzx⟩ ⟨w, hwx⟩ hyz hzw
+
+protected theorem mem (hx : x.IsOrdinal) (hy : y ∈ x) : y.IsOrdinal := by
+  have := hx.isTrans
+  let f : Subrel (· ∈ ·) y.toSet ↪r Subrel (· ∈ ·) x.toSet :=
+    Subrel.inclusionEmbedding (· ∈ ·) (hx.subset_of_mem hy)
+  exact isOrdinal_iff_isTrans.2 ⟨fun _ hz _ ha ↦ hx.mem_trans' ha hz hy, f.isTrans⟩
```
Comment (alreadydone) [inline Mathlib/SetTheory/ZFC/Ordinal.lean:138]:
```suggestion
protected theorem mem (hx : x.IsOrdinal) (hy : y ∈ x) : y.IsOrdinal :=
  have := hx.isTrans
  let f : Subrel (· ∈ ·) y.toSet ↪r Subrel (· ∈ ·) x.toSet :=
    Subrel.inclusionEmbedding (· ∈ ·) (hx.subset_of_mem hy)
  isOrdinal_iff_isTrans.2 ⟨fun _ hz _ ha ↦ hx.mem_trans' ha hz hy, f.isTrans⟩
```

File: Mathlib/SetTheory/ZFC/Ordinal.lean [hunk 3]
```diff
@@ -120,17 +120,130 @@ theorem mem_trans (h : z.IsOrdinal) : x ∈ y → y ∈ z → x ∈ z :=
...
+
+theorem not_mem_iff_subset (hx : x.IsOrdinal) (hy : y.IsOrdinal) : x ∉ y ↔ y ⊆ x := by
+  refine ⟨?_, fun hxy hyx ↦ mem_irrefl _ (hxy hyx)⟩
+  revert hx hy
+  apply Sym2.GameAdd.induction mem_wf _ x y
+  intros x y IH hx hy hyx z hzy
+  by_contra hzx
+  exact hyx (mem_of_subset_of_mem hx hy (IH z x (Sym2.GameAdd.fst_snd hzy) (hy.mem hzy) hx hzx) hzy)
+
+theorem not_subset_iff_mem (hx : x.IsOrdinal) (hy : y.IsOrdinal) : ¬ x ⊆ y ↔ y ∈ x := by
```
Comment (alreadydone) [inline Mathlib/SetTheory/ZFC/Ordinal.lean:189]: Maybe we should define `instance : HasSSubset ZFSet` so you can state that ∈ is the same as ⊂.

**Review by alreadydone — APPROVED** [review]
Thanks!
maintainer delegate

File: Mathlib/SetTheory/ZFC/Ordinal.lean (hunk 1 above)
Comment (vihdzp) [reply Mathlib/SetTheory/ZFC/Ordinal.lean:21]: Oh alright, I'll revert.

File: Mathlib/SetTheory/ZFC/Ordinal.lean (hunk 3 above)
Comment (vihdzp) [reply Mathlib/SetTheory/ZFC/Ordinal.lean:190]: I like that idea. We're missing a few instances on `PSet` and `ZFSet` - most notably, we don't have the `Preorder` / `PartialOrder` structure defined. I'll take care of that in another PR.

File: Mathlib/SetTheory/ZFC/Ordinal.lean [hunk 4]
```diff
@@ -120,17 +121,130 @@ theorem mem_trans (h : z.IsOrdinal) : x ∈ y → y ∈ z → x ∈ z :=
...
+    x.IsOrdinal ↔ x.IsTransitive ∧ IsTrans x.toSet (Subrel (· ∈ ·) _) where
+  mp h := ⟨h.isTransitive, h.isTrans⟩
+  mpr := by
+    rintro ⟨h₁, ⟨h₂⟩⟩
+    refine ⟨h₁, @fun y z w hyz hzw hwx ↦ ?_⟩
+    have hzx := h₁.mem_trans hzw hwx
+    exact h₂ ⟨y, h₁.mem_trans hyz hzx⟩ ⟨z, hzx⟩ ⟨w, hwx⟩ hyz hzw
+
+protected theorem mem (hx : x.IsOrdinal) (hy : y ∈ x) : y.IsOrdinal :=
+  have := hx.isTrans
```
Comment (b-mehta) [inline Mathlib/SetTheory/ZFC/Ordinal.lean:136]: what's this needed for?

**Review by b-mehta — COMMENTED** [review]
bors d+

File: Mathlib/SetTheory/ZFC/Ordinal.lean (hunk 4 above)
Comment (vihdzp) [reply Mathlib/SetTheory/ZFC/Ordinal.lean:136]: Instance argument for `f.isTrans` in the last line.

**Pull request closed without merging**